presentation/src/main/java/com/anytypeio/anytype/presentation/multiplayer/SpaceJoinRequestViewModel.kt

```diff
@@ -155,147 +111,223 @@ class SpaceJoinRequestViewModel(
+            else -> createOtherState(spaceView, spaceMembers, newMember, isCurrentUserOwner)
+        }
+        _viewState.value = state
+    }
+
+    private fun createExplorerState(
+        spaceView: ObjectWrapper.SpaceView,
+        spaceMembers: List<ObjectWrapper.SpaceMember>,
+        newMember: ObjectWrapper.SpaceMember?,
+        isCurrentUserOwner: Boolean
+    ): ViewState {
+        val canAddReaders = spaceView.canAddReaders(isCurrentUserOwner, spaceMembers)
+        val canAddWriters = spaceView.canAddWriters(isCurrentUserOwner, spaceMembers)
+
+        return when {
+            newMember == null -> ViewState.Error.EmptyMember
+            !canAddReaders && !canAddWriters -> createViewStateWithButtons(spaceView, newMember, listOf(InviteButton.UPGRADE, InviteButton.REJECT))
+            else -> createViewStateWithButtons(
+                spaceView,
+                newMember,
+                buildList {
+                    if (!canAddReaders) add(InviteButton.JOIN_AS_VIEWER_DISABLED) else add(InviteButton.JOIN_AS_VIEWER)
+                    if (!canAddWriters) add(InviteButton.JOIN_AS_EDITOR_DISABLED) else add(InviteButton.JOIN_AS_EDITOR)
+                    add(InviteButton.REJECT)
+                }
+            )
+        }
+    }
+
+    private fun createBuilderState(
+        spaceView: ObjectWrapper.SpaceView,
+        spaceMembers: List<ObjectWrapper.SpaceMember>,
+        newMember: ObjectWrapper.SpaceMember?,
+        isCurrentUserOwner: Boolean
+    ): ViewState {
+        val canAddReaders = spaceView.canAddReaders(isCurrentUserOwner, spaceMembers)
+        val canAddWriters = spaceView.canAddWriters(isCurrentUserOwner, spaceMembers)
+        return when {
+            newMember == null -> ViewState.Error.EmptyMember
+            !canAddReaders && !canAddWriters -> createViewStateWithButtons(spaceView, newMember, listOf(InviteButton.ADD_MORE_VIEWERS, InviteButton.ADD_MORE_EDITORS, InviteButton.REJECT))
+            else -> createViewStateWithButtons(
+                spaceView,
+                newMember,
+                buildList {
+                    if (!canAddReaders) add(InviteButton.ADD_MORE_VIEWERS) else add(InviteButton.JOIN_AS_VIEWER)
+                    if (!canAddWriters) add(InviteButton.ADD_MORE_EDITORS) else add(InviteButton.JOIN_AS_EDITOR)
+                    add(InviteButton.REJECT)
+                }
+            )
+        }
+    }
+
+    private fun createOtherState(
+        spaceView: ObjectWrapper.SpaceView,
+        spaceMembers: List<ObjectWrapper.SpaceMember>,
+        newMember: ObjectWrapper.SpaceMember?,
+        isCurrentUserOwner: Boolean
+    ): ViewState {
+        val canAddReaders = spaceView.canAddReaders(isCurrentUserOwner, spaceMembers)
+        val canAddWriters = spaceView.canAddWriters(isCurrentUserOwner, spaceMembers)
+        return when {
+            newMember == null -> ViewState.Error.EmptyMember
```

Comment:
please add error handling or remove this

Reply:
done